channel/test/adjudicator.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package test
+
+import (
+	"context"
+	"math/rand"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"perun.network/go-perun/channel"
+	pkgtest "perun.network/go-perun/pkg/test"
+)
+
+// Adjudicator represents an adjudicator test setup.
+//
+// Adjudicator should return the adjudicator instance used for testing.
+// NewRegisterReq should return the inputs to a successful register call.
+// NewProgressReq should return the inputs to a successful progress call.
+// NewWithdrawReq should return the inputs to a successful withdraw call.
+type Adjudicator interface {
+	Adjudicator() channel.Adjudicator
+	NewRegisterReq(context.Context, *rand.Rand) (*channel.AdjudicatorReq, []channel.SignedState)
+	NewProgressReq(context.Context, *rand.Rand) *channel.ProgressReq
+	NewWithdrawReq(context.Context, *rand.Rand) (*channel.AdjudicatorReq, channel.StateMap)
+}
+
+// TestAdjudicator tests the given adjudicator.
+func TestAdjudicator(ctx context.Context, t *testing.T, a Adjudicator) {
+	rng := pkgtest.Prng(t)
+
+	adj := a.Adjudicator()
+
+	t.Run("register and subscribe", func(t *testing.T) {
+		req, subChannels := a.NewRegisterReq(ctx, rng)
+		err := adj.Register(ctx, *req, subChannels)
```

Comment:
If the framework had a way to explicitly query the chain with eg. a `DisputeReader`, then it could verify that this worked.

Reply:
What exactly would you check? How would the implementation look like? I think how disputes are registered is an implementation detail. I can't think of a way currently how to abstract that across all backends. Probably would need a new interface for that, but usually we check the state after registering with a subscription, not via a call to a disputeReader interface.